src/ddscxx/include/dds/topic/detail/ContentFilteredTopic.hpp

```diff
@@ -28,43 +28,224 @@
+  static constexpr std::size_t Arity = sizeof...(Args);
+
+  template < std::size_t N, std::enable_if_t<N<Arity> * = nullptr>
+  struct argument
+  {
+    using type = typename std::tuple_element<N, std::tuple<Args...>>::type;
+  };
+
+  using argument_tuple = std::tuple<Args...>;
+};
+
+// function pointer
+template<class R, class ... Args>
+struct callable_traits<R (*)(Args...)>
+  : public callable_traits<R(Args...)> {};
+
+// member function pointer
+template<class C, class R, class ... Args>
+struct callable_traits<R (C::*)(Args...)>
+  : public callable_traits<R(C &, Args...)> {};
+
+// const member function pointer
+template<class C, class R, class ... Args>
+struct callable_traits<R (C::*)(Args...) const>
+  : public callable_traits<R(C &, Args...)> {};
+
+// member object pointer
+template<class C, class R>
+struct callable_traits<R(C::*)>
+  : public callable_traits<R(C &)> {};
+
+// functor
+template<class F>
+struct callable_traits
+{
+private:
+  using call_type = callable_traits<decltype(&F::operator())>;
+
+public:
+  using return_type = typename call_type::return_type;
+
+  static constexpr std::size_t Arity = call_type::Arity - 1;
+
+  template < std::size_t N, std::enable_if_t<N<Arity> * = nullptr>
+  struct argument
+  {
+    using type = typename call_type::template argument<N + 1>::type;
+  };
+
+  using argument_tuple = remove_first_from_tuple_t<typename call_type::argument_tuple>;
+};
+
+template<class F>
+struct callable_traits<F &>
+  : public callable_traits<F> {};
+
+template<class F>
+struct callable_traits<F &&>
+  : public callable_traits<F> {};
+
+// Concepts of possible callbacks
+template<class F, class T>
+using sample_only_t = std::enable_if_t<callable_traits<std::decay_t<F>>::Arity == 1 &&
+    std::is_same<typename callable_traits<std::decay_t<F>>::template argument<0>::type,
+    const T &>::value &&
+    std::is_same<typename callable_traits<std::decay_t<F>>::return_type, bool>::value
+>;
+
+template<class F>
+using sample_info_only_t = std::enable_if_t<callable_traits<std::decay_t<F>>::Arity == 1 &&
+    std::is_same<typename callable_traits<std::decay_t<F>>::template argument<0>::type,
+    const dds::sub::SampleInfo &>::value &&
+    std::is_same<typename callable_traits<std::decay_t<F>>::return_type, bool>::value
+>;
+
+template<class F, class T>
+using sample_and_sample_info_t = std::enable_if_t<callable_traits<std::decay_t<F>>::Arity == 2 &&
+    std::is_same<typename callable_traits<std::decay_t<F>>::template argument<0>::type,
+    const T &>::value &&
+    std::is_same<typename callable_traits<std::decay_t<F>>::template argument<1>::type,
+    const dds::sub::SampleInfo &>::value &&
+    std::is_same<typename callable_traits<std::decay_t<F>>::return_type, bool>::value
+>;
+
+} // namespace detail
+
+
 namespace dds {
 namespace topic {
 namespace detail {
 
 class FunctorHolderBase
 {
 public:
-    FunctorHolderBase() { };
-
-    virtual ~FunctorHolderBase() { };
+    FunctorHolderBase() = default;
```

Comment:
I think all these explicitly defined default copy/move constructors/assignment operators are superfluous, since when you define none of them the compiler will generate them implicitly for you.